aomi/model/__init__.py

```diff
@@ -0,0 +1,626 @@
+    are actually available for use. Uses tags and command line
+    options to make determination."""
+
+    ctx = Context(context.opt)
+    for resource in context.resources():
+        if resource.child:
+            continue
+        if resource.filtered():
+            ctx.add(resource)
+
+    return ctx
+
+
+class Context(object):
+    """The overall context of an aomi session"""
+
+    @staticmethod
+    def load(config, opt):
+        """Loads and returns a full context object based on the Secretfile"""
+        ctx = Context(opt)
+        seed_map = py_resources()
+        seed_keys = set([m[0] for m in seed_map])
+        for config_key in seed_keys:
+            if config_key not in config:
+                continue
+            for resource in config[config_key]:
+                mod = find_model(config_key, resource, seed_map)
+                if not mod:
+                    print("unable to find mod for %s" % resource)
+                    continue
+
+                ctx.add(mod(resource, opt))
+
+        for config_key in config.keys():
+            if config_key != 'pgp_keys' and \
+               config_key not in seed_keys:
+                print("missing model for %s" % config_key)
+
+        f_ctx = aomi.model.filtered_context(ctx)
+        return f_ctx
+
+    def thaw(self, tmp_dir):
+        """Will thaw a secret into an appropriate location"""
+        for resource in self.resources():
+            if resource.present:
+                resource.thaw(tmp_dir)
+
+    def freeze(self, dest_dir):
+        """Freezes everyt resource within a context"""
+        for resource in self.resources():
+            if resource.present:
+                resource.freeze(dest_dir)
+
+    def __init__(self, opt):
+        self._mounts = []
+        self._resources = []
+        self._auths = []
+        self._logs = []
+        self.opt = opt
+
+    def mounts(self):
+        """Secret backends within context"""
+        return self._mounts
+
+    def logs(self):
+        """Audit log backends within context"""
+        return self._logs
+
+    def auths(self):
+        """Authentication backends within context"""
+        return self._auths
+
+    def resources(self):
+        """Vault resources within context"""
+        res = []
+        for resource in self._resources:
+            res = res + resource.resources()
+
+        return res
+
+    def add(self, resource):
+        """Add a resource to the context"""
+        if isinstance(resource, Resource):
+            if isinstance(resource, (Secret, Mount)):
+                ensure_backend(resource, SecretBackend, self._mounts, self.opt)
+            elif isinstance(resource, (Auth)):
+                ensure_backend(resource, AuthBackend, self._auths, self.opt)
+            elif isinstance(resource, (AuditLog)):
+                ensure_backend(resource, LogBackend, self._logs, self.opt)
+
+            self._resources.append(resource)
```

Comment:
Same comment as above... We aren't chaining `add` anywhere, it's executed on an object-by-object basis within an iterator.